### PR TITLE
Use `assert_dom_equal` more in Action View Tests

### PR DESCRIPTION
**Pull request by seanpdoyle**
### Motivation / Background

To [quote @-flavorjones][comment]

> asserting against string constants is error-prone and brittle, and it
> would be better to have a testing framework that asserted on semantic
> equivalence

### Detail

When appropriate, use [assert_dom_equal][] instead of `assert_equal` for String comparisons of Action View Helper-constructed values.

Some exceptions to that guideline are tests that assert equivalence of:

* instances of Objects
* URLs
* view partial paths
* String values not intended for HTML documents

[comment]: https://github.com/seanpdoyle/rails/pull/1#issuecomment-799656813
[assert_dom_equal]: https://github.com/rails/rails-dom-testing#dom-assertions

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by seanpdoyle** [comment]
@flavorjones I'm in the process of re-attempting https://github.com/rails/rails/pull/41656. I'm proposing this changeset based on a [comment you left](https://github.com/seanpdoyle/rails/pull/1#issuecomment-799656813).

I'd really appreciate hearing more on your opinion and, and a review of these changes.

**Comment by flavorjones** [comment]
@seanpdoyle Thanks for doing this work! As you pointed out, I am generally in favor of using `assert_dom_equal` more often, and so I am enthusastic that this is taking the project in the right direction.

I don't have any strong objections to these changes -- especially since the tests are green! But I want to point out that there are a couple of cases that maybe _should_ be using `assert_equals`. To help frame a deeper conversation about guiding principles, I'll provide a few examples from the PR ...

This should absolutely be using `assert_dom_equal`:

```ruby
assert_dom_equal "<meta property=\"csp-nonce\" name=\"csp-nonce\" content=\"iyhD0Yc0W+c=\" />", csp_meta_tag(property: "csp-nonce")
```

My reasoning is that the method `csp_meta_tag` could be changed to emit slightly different bytes (e.g., whitespace) that would still be DOM-equivalent, and this test should still pass. ✅ 

This one should probably _not_ be using `assert_dom_equal`:

```ruby
assert_dom_equal "-111.235", number_with_precision(-111.2346)
```

My reasoning is that we're testing for a specific output string, and not checking for equivalent DOM structure. A good test for this might be, "Could I use this method to generate an SMS text", and if the answer is "yes" then we should probably avoid `assert_dom_equals` and prefer `assert_equals`.

And this one is solidly in the middle -- I think it's a judgement call whether you think we're testing for DOM equivalence or these specific bytes:

```ruby
  def test_existing_nested_resource
    @controller = SessionsController.new

    get :show, params: { workshop_id: 1, id: 1 }
    assert_dom_equal %{/workshops/1/sessions/1\n<a href="/workshops/1/sessions/1">Session</a>}, @response.body
  end
```

**TLDR** maybe some guiding principles I'd suggest we adopt (and that maybe guide a second pass through this PR):

- any string that is testing DOM structure or HTML entities and escaping should absolutely use `assert_dom_equals`
- any string that is being tested outside of an HTML context should use `assert_equals`
- for tests that aren't clearly in one camp or the other, default to `assert_dom_equals`

WDYT? Is this helpful feedback?

**Comment by seanpdoyle** [comment]
>  Is this helpful feedback?

It is! I had originally favored fewer changes in the hopes of making the proposal less challenging to review. After that, my gut feeling shifted to _over_ testing, since two DOM Strings without structure are equivalent, and missing a DOM test would be more challenging to catch.

I'm certainly willing to revert several changes along the guidelines you've outlined to strike a compromise.

**Comment by rafaelfranca** [comment]
> My reasoning is that the method csp_meta_tag could be changed to emit slightly different bytes (e.g., whitespace) that would still be DOM-equivalent, and this test should still pass. ✅

That is a reason to not use those methods. We don't want to be generating extra whitespaces. We don't only want to generate equivalent dom, we want to make sure the dom we generate is clean.

A string like `"<meta      property=\"csp-nonce\" name=\"csp-nonce\" content=\"iyhD0Yc0W+c=\" />"` should not pass that test.

**Comment by seanpdoyle** [comment]
@rafaelfranca thank you for that feedback. It sounds like replacing `assert_equal` with `assert_dom_equal` isn't something we're interested in doing globally.

Are there changes in this diff that are good candidates, or would it be best to close?

**Comment by rafaelfranca** [comment]
Hard to tell, diff is too big. I'd say that if the assertion is testing the dom, but not really the result of HTML generated by Rails, it is fine to change. If it is testing the result of HTML generated by Rails, it needs to be as precise as possible. I don't want Rails to start generating garbage HTML and our tests not catching it

**Comment by flavorjones** [comment]
Thanks for clarifying what you think the guiding principles should be, @rafaelfranca! This was really helpful for me to read, I've updated my mental models.

**Comment by seanpdoyle** [comment]
Thank you for this discussion. I'm going to close this for now and consider the feedback.
